### PR TITLE
Apply the #1514 fix even on non-Windows platforms, and change all remaining cudaStreamSynchronize calls

### PR DESCRIPTION
**Pull request by reunanen**
As [it seems](https://github.com/davisking/dlib/issues/1513#issuecomment-447716981) that #1513 may occur also outside Windows (e.g., RHEL 7), make the #1514 fix apply on other platforms as well.

### CODE REVIEW

**Comment by davemers0160** [comment]
@reunanen : These are the changes that I have been running with on the RHEL 7 side and I still get the occasional freeze.  I'm only seeing it happen in about 1 in 20 training runs.  I think that the multilple GPU's (>2) may also exagerate the problem.  I can recompile for 2 GPUs vs. 4 and see if I can get the freeze to happen.

**Comment by reunanen** [comment]
@davemers0160 Oh, ok, thanks for the update.
Are you sure it freezes inside `synchronize_stream`? (If not, have you tried to replace the remaining `cudaStreamSynchronize(0)` calls -- along [these lines](https://github.com/reunanen/dlib/commit/9bc39ef04f5deae27da351ae3ef7f45cafa6ef20)?)

**Comment by davemers0160** [comment]
@reunanen : I'm not 100% sure.  I'll try to add some simple cout statements to each of the case statements to see where things freeze.

**Comment by davisking** [comment]
Cool, thanks for another PR.  I'll wait a bit for everyone to report back about this fixing or not fixing it before merging.  I'm also tempted to say we should put an #ifdef around it that makes it active only for specific CUDA versions, like maybe just for 10?  Since surely NVIDIA will fix the issue and polling is somewhat of an expensive thing to do.  So it would be nice to eventually use the blocking API.

**Comment by reunanen** [comment]
@davemers0160 Ok, great. I just hope the cout statements won't essentially change the relative timings too much (Heisenbug).

@davisking Completely agree. As per [this information](https://github.com/davisking/dlib/issues/1513#issuecomment-437709703), made it active on CUDA 9.2 and 10.0. Although [apparently](https://github.com/davisking/dlib/issues/1513#issuecomment-444094629) 9.0 may be affected as well.

File: dlib/cuda/gpu_data.cpp [hunk 1]
```diff
@@ -79,12 +79,33 @@ namespace dlib
...
+            cudaError_t err = cudaStreamQuery(stream);
+            switch (err)
+            {
+            case cudaSuccess: return;      // now we are synchronized
+            case cudaErrorNotReady: break; // continue waiting
+            default: CHECK_CUDA(err);      // unexpected error: throw
+            }
+        }
+#else // CUDA_VERSION
+        cudaStreamSynchronize(stream);
```
[the remark this answers is not in the document]
Comment (reunanen) [reply dlib/cuda/gpu_data.cpp]: Oops, fixed.

**Comment by davemers0160** [comment]
@reunanen : I think you're right about Heisenburg.  I think the cout's are slowing thing down just enough.  I even added a counter to spit something out every 65535 ticks.   I can't get anything to freeze.  I can definitely say that the while loop in the 'synchronize_stream' function sits in the 'cudaErrorNotReady' case for a while before the 'cudaSuccess' is hit.  It appears that the training sits in the not ready case for a shorter amount of time at the beginning and then increases as training goes on.

@davisking : I've seen the freeze happen on the RHEL7 system with cuda 9.0.  I would recommend going back at least that far.

**Comment by reunanen** [comment]
@davemers0160 Ok, dammit. I know this is more work, but could you think of adding an atomic variable which is for example incremented once every time `cudaStreamQuery` is called (this incrementing update should be a very lightweight operation compared to cout output), and then a new thread which only outputs the value of this variable every, let's say, 10 minutes? Alternatively, the variable could contain the latest [timepoint](https://en.cppreference.com/w/cpp/chrono/steady_clock/now) when `cudaStreamQuery` was called, or something like that. See what I mean?

**Comment by davisking** [comment]
@davemers0160 What hardware were you using when you experienced the problem in RHEL with cuda 9.0?  I've never seen this issue myself.  I've only got a 1080ti and a titan x though.

**Comment by davemers0160** [comment]
@reunanen : I think that I've got something working.  Fingers crossed and hopefully I'm not slowing 'synchronize_stream' too bad.

@davisking  :  Four Tesla P100-SXM2-16GB, but I have seen it happen on dual 1080Ti system as well.

**Comment by davemers0160** [comment]
OK.  I got it.  I was able to get several training runs to stall at various times during the training.  Here's the modified code (nr and nr2 are passed to another thread and sent to std out every 300 seconds):

```
void synchronize_stream(cudaStream_t stream)
{
  while (true)
  {
    tp = (std::chrono::steady_clock::now()).time_since_epoch();
    nr2 = tp.count();
    cudaError_t err = cudaStreamQuery(stream);
    switch (err)
    {
      case cudaSuccess:    return;
      case cudaErrorNotReady:
        ++nr;
        break;
      default: CHECK_CUDA(err);
    }
  }
} 
```

In each case where the freeze occurred both nr and nr2 stop updating.  To me that would mean that I'm stuck at the `cudaStreamQuery(stream)` line.  Since NULL is being passed to the function does anyone know how it handles multiple GPUs (I'm assuming that a new stream is created for each GPU).

**Comment by davisking** [comment]
Hmm, that is disappointing.  Multiple GPUs should not matter, the cuda API should only be interacting with the currently selected GPU, the one indicated by `cudaGetDevice`.

**Comment by davemers0160** [comment]
I'm not sure where we go from here.  I'm not as well versed on the cuda side of things, but this could be just extreme corner cases since just @reunanen and I have seen it to date.

**Comment by davisking** [comment]
It is sort of surprising.  This seems like a serious bug in cuda.  Have either of you reported it to nvidia?

Also, `dlib/CMakeLists.txt` sets `-arch=sm_30`.  Maybe setting it to a newer arch would make the problem go away?  `sm_30` is somewhat dated, maybe newer cuda runtimes are less tested with that setting.  It might be time to update that to a higher requirement, although that should probably be conditional on the version of cuda found.

**Comment by davemers0160** [comment]
@davisking : I tried changing the arch setting to `-arch=sm_60` and still get the freeze.  Right now I'm trying the same architecture using only 1 and 2 GPUs to see if it is related to the number of GPUs used during training.

**Comment by davisking** [comment]
Dang :(

**Comment by davemers0160** [comment]
@davisking : Well, everything finished up and it looks like this only occurs when using multiple GPUs.  So far the freeze has occurred when I use 2 and 4 GPUs but not with one GPU.

**Comment by davisking** [comment]
Huh.  I think @reunanen encountered the issue when using just one GPU though right?

**Comment by reunanen** [comment]
@davisking Correct. But I haven't been able to reproduce it after #1514. Before that, it occurred on RTX 2080 Ti, but for me never on GTX 1080 Ti (tested on dual rigs too).

**Comment by davisking** [comment]
So what's the state of this whole cuda stream deadlocking issue?  It seems like this PR should be merged as it mitigates *some* of the issue, while not totally avoiding all deadlocks?

**Comment by davemers0160** [comment]
I think that it is worth merging.  So far it has fixed everything except my small test case.

**Comment by reunanen** [comment]
So this #1596 does appear to help in some specific case(s) that #1514 did not fix already?

(I have not seen any – #1514 fixed the only freezing issue I ever had.)

**Comment by davisking** [comment]
Sounds good to me.  Thanks guys for debugging into this.  Hopefully the next version of cuda removes the issue.

**Comment by reunanen** [comment]
> Hopefully the next version of cuda removes the issue.

For what it's worth: at least I have been on CUDA 11 for a long time already, and haven't noticed this problem.

**Comment by davisking** [comment]
Na `get_net()` definitely shouldn't be needed.  I wager mightily it's just pushing a race condition in the cuda runtime around and making it less likely to happen.

**Pull request merged**